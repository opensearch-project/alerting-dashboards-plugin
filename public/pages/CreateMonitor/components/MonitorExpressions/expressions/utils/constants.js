/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const POPOVER_STYLE = { zIndex: '200' };
export const EXPRESSION_STYLE = { padding: '20px', whiteSpace: 'nowrap' };
export const Expressions = {
  THRESHOLD: 'THRESHOLD',
  WHEN: 'WHEN',
  OF_FIELD: 'OF_FIELD',
  OVER: 'OVER',
  FOR_THE_LAST: 'FOR_THE_LAST',
  WHERE: 'WHERE',
  METRICS: 'METRICS',
  GROUP_BY: 'GROUP_BY',
};
export const NUMBER_TYPES = [
  'long',
  'integer',
  'short',
  'byte',
  'double',
  'float',
  'half_float',
  'scaled_float',
];
export const UNITS_OF_TIME = [
  { value: 'm', text: 'minute(s)' },
  { value: 'h', text: 'hour(s)' },
  { value: 'd', text: 'day(s)' },
];

export const WHERE_BOOLEAN_FILTERS = [
  { text: 'Select value', value: '' },
  { text: 'True', value: true },
  { text: 'False', value: false },
];

export const OPERATORS_MAP = {
  IS: {
    text: 'is',
    value: 'is',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT: {
    text: 'is not',
    value: 'is_not',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NULL: {
    text: 'is null',
    value: 'is_null',
    dataTypes: ['number', 'text', 'keyword', 'boolean'],
  },
  IS_NOT_NULL: {
    text: 'is not null',
    value: 'is_not_null',
    dataTypes: ['number', 'text', 'keyword'],
  },
  IS_GREATER: {
    text: 'is greater than',
    value: 'is_greater',
    dataTypes: ['number'],
  },
  IS_GREATER_EQUAL: {
    text: 'is greater than equal',
    value: 'is_greater_equal',
    dataTypes: ['number'],
  },
  IS_LESS: {
    text: 'is less than',
    value: 'is_less',
    dataTypes: ['number'],
  },
  IS_LESS_EQUAL: {
    text: 'is less than equal',
    value: 'is_less_equal',
    dataTypes: ['number'],
  },
  STARTS_WITH: {
    text: 'starts with',
    value: 'starts_with',
    dataTypes: ['text', 'keyword'],
  },
  ENDS_WITH: {
    text: 'ends with',
    value: 'ends_with',
    dataTypes: ['text', 'keyword'],
  },
  CONTAINS: {
    text: 'contains',
    value: 'contains',
    dataTypes: ['text', 'keyword'],
  },
  DOES_NOT_CONTAINS: {
    text: 'does not contain',
    value: 'does_not_contains',
    dataTypes: ['text'],
  },
  IN_RANGE: {
    text: 'is in range',
    value: 'in_range',
    dataTypes: ['number'],
  },
  NOT_IN_RANGE: {
    text: 'is not in range',
    value: 'not_in_range',
    dataTypes: ['number'],
  },
};

export const OVER_TYPES = [{ value: 'all documents', text: 'all documents' }];

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'count', text: 'count()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const GROUP_BY_ERROR = 'Must specify at least 1 group by expression.';
export const QUERY_TYPE_GROUP_BY_ERROR = 'Can have a maximum of 1 group by selections.';

export const QUERY_TYPE_METRIC_ERROR = 'Can have a maximum of 1 metric selections.';

export const MAX_NUM_QUERY_LEVEL_GROUP_BYS = 1;
export const MAX_NUM_BUCKET_LEVEL_GROUP_BYS = 2;
export const MAX_NUM_WHERE_EXPRESSION = 1;
export const WHERE_FILTER_ALLOWED_TYPES = ['number', 'text', 'keyword', 'boolean'];
