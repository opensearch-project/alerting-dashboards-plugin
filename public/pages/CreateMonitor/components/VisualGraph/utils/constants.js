/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const SIZE_RANGE = [3, 3];
export const LINE_STYLES = { strokeWidth: 1 };
export const ANNOTATION_STYLES = { strokeWidth: 1, stroke: 'red' };
export const HINT_STYLES = {
  background: '#3a3a48',
  borderRadius: '5px',
  padding: '7px 10px',
  fontSize: '12px',
  fontWeight: 'bold',
  color: '#fff',
  textAlign: 'left',
  whiteSpace: 'nowrap',
  boxShadow: '0 2px 4px rgba(0,0,0,.5)',
};

// Don't want values to reach top of graph so increase by a buffer % so we always have a yDomain with a slightly higher max
export const Y_DOMAIN_BUFFER = 1.4; // 40%

export const X_DOMAIN_BUFFER = 0.2; // 40%

// Size of circle for each point on graph
export const DEFAULT_MARK_SIZE = 3;

export const BAR_PERCENTAGE = 0.015;

export const BAR_KEY_COUNT = 3;
