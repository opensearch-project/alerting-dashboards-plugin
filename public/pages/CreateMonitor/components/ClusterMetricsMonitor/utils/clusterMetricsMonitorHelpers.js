/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import _ from 'lodash';
import { formikToClusterMetricsInput } from '../../../containers/CreateMonitor/utils/formikToMonitor';
import {
  API_TYPES,
  GET_API_TYPE_DEBUG_TEXT,
  ILLEGAL_PATH_PARAMETER_CHARACTERS,
  NO_PATH_PARAMS_PLACEHOLDER_TEXT,
  PATH_PARAMETER_ILLEGAL_CHARACTER_TEXT,
  PATH_PARAMETERS_REQUIRED_TEXT,
} from './clusterMetricsMonitorConstants';
import { FORMIK_INITIAL_VALUES } from '../../../containers/CreateMonitor/utils/constants';

export function buildClusterMetricsRequest(values) {
  return _.get(formikToClusterMetricsInput(values), 'uri');
}

export const canExecuteClusterMetricsMonitor = (uri = {}) => {
  const {
    api_type = FORMIK_INITIAL_VALUES.uri.api_type,
    path_params = FORMIK_INITIAL_VALUES.uri.path_params,
  } = uri;
  if (_.isEmpty(api_type)) return false;
  const requiresPathParams = _.isEmpty(_.get(API_TYPES, `${api_type}.paths.withoutPathParams`));
  const hasPathParams = !_.isEmpty(path_params);
  if (pathParamsContainIllegalCharacters(path_params)) return false;
  return requiresPathParams ? hasPathParams : true;
};

export const pathParamsContainIllegalCharacters = (pathParams) => {
  if (_.isEmpty(pathParams)) return false;
  const foundIllegalCharacters = ILLEGAL_PATH_PARAMETER_CHARACTERS.find((illegalCharacter) =>
    _.includes(pathParams, illegalCharacter)
  );
  return !_.isEmpty(foundIllegalCharacters);
};

export const getApiPath = (hasPathParams = false, apiType) => {
  let path = hasPathParams
    ? _.get(
        API_TYPES,
        `${apiType}.paths.withPathParams`,
        _.get(API_TYPES, `${apiType}.paths.withoutPathParams`)
      )
    : _.get(API_TYPES, `${apiType}.paths.withoutPathParams`);
  return path || FORMIK_INITIAL_VALUES.uri.path;
};

export const getApiTypesRequiringPathParams = () => {
  const apiList = [];
  _.keys(API_TYPES).forEach((api) => {
    const withoutPathParams = _.get(API_TYPES, `${api}.paths.withoutPathParams`, '');
    if (_.isEmpty(withoutPathParams))
      apiList.push({
        value: _.get(API_TYPES, `${api}.type`),
        label: _.get(API_TYPES, `${api}.label`),
      });
  });
  return apiList;
};

export const getExamplePathParams = (apiType) => {
  if (_.isEmpty(apiType)) return NO_PATH_PARAMS_PLACEHOLDER_TEXT;
  let exampleText = _.get(API_TYPES, `${apiType}.exampleText`, '');
  _.isEmpty(exampleText)
    ? (exampleText = NO_PATH_PARAMS_PLACEHOLDER_TEXT)
    : (exampleText = `e.g., ${exampleText}`);
  return exampleText;
};

export const getApiType = (uri) => {
  let apiType = '';
  // Trim '/' characters from the beginning and end of the path
  const path = _.get(uri, 'path')?.replace(/^\/+|\/+$/g, '');
  if (_.isEmpty(path)) return apiType;
  _.keys(API_TYPES).forEach((apiTypeKey) => {
    const withPathParams = _.get(API_TYPES, `${apiTypeKey}.paths.withPathParams`);
    const withoutPathParams = _.get(API_TYPES, `${apiTypeKey}.paths.withoutPathParams`);
    if (path === withPathParams || path === withoutPathParams) apiType = apiTypeKey;
  });
  if (_.isEmpty(apiType)) console.debug(GET_API_TYPE_DEBUG_TEXT);
  return apiType;
};

export const isInvalidApiPathParameter = (field, hidePathParams, pathParams, requirePathParams) => {
  if (hidePathParams) return false;
  const pathParamsTouched = _.get(field, 'touched.uri.path_params', false);
  if (pathParamsTouched) {
    if (requirePathParams && _.isEmpty(pathParams)) return true;
    return pathParamsContainIllegalCharacters(pathParams);
  } else return pathParamsTouched;
};

export const validateApiPathParameter = (field, hidePathParams, pathParams, requirePathParams) => {
  if (hidePathParams) return NO_PATH_PARAMS_PLACEHOLDER_TEXT;
  const pathParamsTouched = _.get(field, 'touched.uri.path_params', false);
  if (requirePathParams && pathParamsTouched && _.isEmpty(pathParams))
    return PATH_PARAMETERS_REQUIRED_TEXT;
  if (isInvalidApiPathParameter(field, hidePathParams, pathParams, requirePathParams))
    return PATH_PARAMETER_ILLEGAL_CHARACTER_TEXT;
};
