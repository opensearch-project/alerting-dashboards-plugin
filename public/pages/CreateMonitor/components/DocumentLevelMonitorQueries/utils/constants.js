/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { OPERATORS_MAP } from '../../MonitorExpressions/expressions/utils/constants';

export const DOC_LEVEL_INPUT_FIELD = 'doc_level_input';

/**
 * A list of the operators currently supported for defining queries through the UI.
 */
export const SUPPORTED_DOC_LEVEL_QUERY_OPERATORS = [
  OPERATORS_MAP.IS.value,
  OPERATORS_MAP.IS_NOT.value,
  OPERATORS_MAP.IS_GREATER.value,
  OPERATORS_MAP.IS_GREATER_EQUAL.value,
  OPERATORS_MAP.IS_LESS.value,
  OPERATORS_MAP.IS_LESS_EQUAL.value,
];

/**
 * Initial implementation of document level monitors used these feature-specific operator values.
 * Refactored document level monitors assets to reuse the OPERATORS_MAP to reduce duplicate code
 * as new query types become supported.
 */
export const LEGACY_QUERY_OPERATORS = {
  IS: { text: 'is', value: '==' },
  IS_NOT: { text: 'is not', value: '!=' },
};

/**
 * These patterns delineate the field being queries from the query value
 * when using query string query syntax.
 */
export const QUERY_STRING_QUERY_OPERATORS = {
  [OPERATORS_MAP.IS_GREATER.value]: ':>',
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: ':>=',
  [OPERATORS_MAP.IS_LESS.value]: ':<',
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: ':<=',
};

/**
 * Similar to OPERATORS_QUERY_MAP, this const contains the query patterns
 * for the SUPPORTED_DOC_LEVEL_QUERY_OPERATORS.
 */
export const DOC_LEVEL_QUERY_MAP = {
  [LEGACY_QUERY_OPERATORS.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [LEGACY_QUERY_OPERATORS.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS.value]: {
    query: ({ field, query }) => `${field}:\"${query}\"`,
  },
  [OPERATORS_MAP.IS_NOT.value]: {
    query: ({ field, query }) => `NOT (${field}:\"${query}\")`,
  },
  [OPERATORS_MAP.IS_GREATER.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER.value]}${query}`,
  },
  [OPERATORS_MAP.IS_GREATER_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_GREATER_EQUAL.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS.value]}${query}`,
  },
  [OPERATORS_MAP.IS_LESS_EQUAL.value]: {
    query: ({ field, query }) =>
      `${field}${QUERY_STRING_QUERY_OPERATORS[OPERATORS_MAP.IS_LESS_EQUAL.value]}${query}`,
  },
};
