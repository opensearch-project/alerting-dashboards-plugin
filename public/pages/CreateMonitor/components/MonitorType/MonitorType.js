/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiFlexGrid, EuiFlexItem, EuiFormLabel, EuiSpacer, EuiText } from '@elastic/eui';
import FormikCheckableCard from '../../../../components/FormControls/FormikCheckableCard';
import { MONITOR_TYPE, SEARCH_TYPE } from '../../../../utils/constants';
import { FORMIK_INITIAL_TRIGGER_VALUES } from '../../../CreateTrigger/containers/CreateTrigger/utils/constants';
import {
  DEFAULT_DOCUMENT_LEVEL_QUERY,
  FORMIK_INITIAL_VALUES,
} from '../../containers/CreateMonitor/utils/constants';

export const MONITOR_TYPE_CARD_WIDTH = 400; // TODO DRAFT: Determine width

const onChangeDefinition = (e, form) => {
  const type = e.target.value;
  form.setFieldValue('monitor_type', type);

  // Clearing various form fields when changing monitor types.
  // TODO: Implement modal that confirms the change before clearing.
  form.setFieldValue('index', FORMIK_INITIAL_VALUES.index);
  form.setFieldValue('searchType', FORMIK_INITIAL_VALUES.searchType);
  form.setFieldValue('triggerDefinitions', FORMIK_INITIAL_TRIGGER_VALUES.triggerConditions);
  switch (type) {
    case MONITOR_TYPE.COMPOSITE_LEVEL:
      form.setFieldValue('searchType', SEARCH_TYPE.GRAPH);
      break;
    case MONITOR_TYPE.CLUSTER_METRICS:
      form.setFieldValue('searchType', SEARCH_TYPE.CLUSTER_METRICS);
      break;
    case MONITOR_TYPE.DOC_LEVEL:
      form.setFieldValue('query', DEFAULT_DOCUMENT_LEVEL_QUERY);
      break;
    default:
      form.setFieldValue('query', FORMIK_INITIAL_VALUES.query);
  }
};

const queryLevelDescription = (
  <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
    Per query monitors run a query and generate alerts based on trigger criteria that match query
    results.
  </EuiText>
);

const bucketLevelDescription = (
  <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
    Per bucket monitors run a query that evaluates trigger criteria based on aggregated values in
    the dataset.
  </EuiText>
);

const clusterMetricsDescription = (
  <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
    Per cluster metrics monitors run API requests to monitor the clusterâ€™s health.
  </EuiText>
);

const documentLevelDescription = // TODO DRAFT: confirm wording
  (
    <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
      Per document monitors run queries that return individual documents matching the trigger
      conditions.
    </EuiText>
  );

const compositeLevelDescription = (
  <EuiText color={'subdued'} size={'xs'} style={{ paddingBottom: '10px', paddingTop: '0px' }}>
    Composite monitors chain the outputs of different monitor types and focus trigger conditions to
    reduce alert noise and generate finer results.
  </EuiText>
);

const MonitorType = ({ values }) => (
  <>
    <EuiFormLabel>Monitor type</EuiFormLabel>
    <EuiSpacer size="xs" />
    <EuiFlexGrid gutterSize={'m'}>
      <EuiFlexItem grow={false} style={{ width: 350 }}>
        <FormikCheckableCard
          name="monitorTypeQueryLevel"
          inputProps={{
            id: 'queryLevelMonitorRadioCard',
            label: 'Per query monitor',
            checked: values.monitor_type === MONITOR_TYPE.QUERY_LEVEL,
            value: MONITOR_TYPE.QUERY_LEVEL,
            onChange: (e, field, form) => onChangeDefinition(e, form),
            children: queryLevelDescription,
            'data-test-subj': 'queryLevelMonitorRadioCard',
          }}
        />
      </EuiFlexItem>
      <EuiFlexItem grow={false} style={{ width: 350 }}>
        <FormikCheckableCard
          name="monitorTypeBucketLevel"
          inputProps={{
            id: 'bucketLevelMonitorRadioCard',
            label: 'Per bucket monitor',
            checked: values.monitor_type === MONITOR_TYPE.BUCKET_LEVEL,
            value: MONITOR_TYPE.BUCKET_LEVEL,
            onChange: (e, field, form) => onChangeDefinition(e, form),
            children: bucketLevelDescription,
            'data-test-subj': 'bucketLevelMonitorRadioCard',
          }}
        />
      </EuiFlexItem>
      <EuiFlexItem grow={false} style={{ width: 350 }}>
        <FormikCheckableCard
          name="monitorTypeClusterMetrics"
          inputProps={{
            id: 'clusterMetricsMonitorRadioCard',
            label: 'Per cluster metrics monitor',
            checked: values.monitor_type === MONITOR_TYPE.CLUSTER_METRICS,
            value: MONITOR_TYPE.CLUSTER_METRICS,
            onChange: (e, field, form) => onChangeDefinition(e, form),
            children: clusterMetricsDescription,
            'data-test-subj': 'clusterMetricsMonitorRadioCard',
          }}
        />
      </EuiFlexItem>
      <EuiFlexItem grow={false} style={{ width: 350 }}>
        <FormikCheckableCard
          name="monitorTypeDocLevel"
          inputProps={{
            id: 'docLevelMonitorRadioCard',
            label: 'Per document monitor',
            checked: values.monitor_type === MONITOR_TYPE.DOC_LEVEL,
            value: MONITOR_TYPE.DOC_LEVEL,
            onChange: (e, field, form) => onChangeDefinition(e, form),
            children: documentLevelDescription,
            'data-test-subj': 'docLevelMonitorRadioCard',
          }}
        />
      </EuiFlexItem>
      <EuiFlexItem grow={false} style={{ width: 350 }}>
        <FormikCheckableCard
          name="monitorTypeCompositeLevel"
          inputProps={{
            id: 'compositeLevelMonitorRadioCard',
            label: 'Composite monitor',
            checked: values.monitor_type === MONITOR_TYPE.COMPOSITE_LEVEL,
            value: MONITOR_TYPE.COMPOSITE_LEVEL,
            onChange: (e, field, form) => onChangeDefinition(e, form),
            children: compositeLevelDescription,
            'data-test-subj': 'compositeLevelMonitorRadioCard',
          }}
        />
      </EuiFlexItem>
    </EuiFlexGrid>
  </>
);

export default MonitorType;
