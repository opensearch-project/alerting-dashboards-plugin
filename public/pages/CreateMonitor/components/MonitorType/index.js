/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import MonitorType from './MonitorType';

export default MonitorType;
