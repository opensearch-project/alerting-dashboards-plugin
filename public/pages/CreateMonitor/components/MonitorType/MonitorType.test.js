/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from 'enzyme';

import MonitorType from './MonitorType';
import { Formik } from 'formik';
import { FORMIK_INITIAL_VALUES } from '../../containers/CreateMonitor/utils/constants';

describe('MonitorType', () => {
  test('renders', () => {
    const component = (
      <Formik initialValues={FORMIK_INITIAL_VALUES}>{() => <MonitorType values={{}} />}</Formik>
    );

    expect(render(component)).toMatchSnapshot();
  });
});
