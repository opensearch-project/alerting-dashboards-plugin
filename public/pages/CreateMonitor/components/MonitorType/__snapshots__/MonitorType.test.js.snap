// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonitorType renders 1`] = `
<div
  class="euiFlexGrid euiFlexGrid--gutterMedium euiFlexGrid--wrap euiFlexGrid--responsive"
>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFormRow"
      id="monitorTypeQueryLevel-form-row-row"
    >
      <div
        class="euiFormRow__labelWrapper"
      >
        <label
          class="euiFormLabel euiFormRow__label"
          for="monitorTypeQueryLevel-form-row"
        >
          Monitor type
        </label>
      </div>
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder euiPanel--flexGrowZero euiSplitPanel euiSplitPanel--row euiCheckableCard"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--subdued euiPanel--noShadow euiPanel--noBorder euiPanel--flexGrowZero euiPanel--isClickable euiSplitPanel__inner"
          >
            <div
              class="euiRadio euiRadio--noLabel"
              data-test-subj="queryLevelMonitorRadioCard"
            >
              <input
                class="euiRadio__input"
                id="queryLevelMonitorRadioCard"
                name="monitorTypeQueryLevel"
                type="radio"
                value="query_level_monitor"
              />
              <div
                class="euiRadio__circle"
              />
            </div>
          </div>
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--transparent euiPanel--noShadow euiPanel--noBorder euiSplitPanel__inner"
          >
            <label
              aria-describedby="queryLevelMonitorRadioCard-details"
              class="euiCheckableCard__label"
              for="queryLevelMonitorRadioCard"
            >
              Per query monitor
            </label>
            <div
              class="euiCheckableCard__children"
              id="queryLevelMonitorRadioCard-details"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding-bottom:10px;padding-top:0px"
              >
                <div
                  class="euiTextColor euiTextColor--subdued"
                >
                  Per query monitors run a query and generate alerts based on trigger criteria that match query results.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFormRow euiFormRow--hasEmptyLabelSpace"
      id="monitorTypeBucketLevel-form-row-row"
    >
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder euiPanel--flexGrowZero euiSplitPanel euiSplitPanel--row euiCheckableCard"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--subdued euiPanel--noShadow euiPanel--noBorder euiPanel--flexGrowZero euiPanel--isClickable euiSplitPanel__inner"
          >
            <div
              class="euiRadio euiRadio--noLabel"
              data-test-subj="bucketLevelMonitorRadioCard"
            >
              <input
                class="euiRadio__input"
                id="bucketLevelMonitorRadioCard"
                name="monitorTypeBucketLevel"
                type="radio"
                value="bucket_level_monitor"
              />
              <div
                class="euiRadio__circle"
              />
            </div>
          </div>
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--transparent euiPanel--noShadow euiPanel--noBorder euiSplitPanel__inner"
          >
            <label
              aria-describedby="bucketLevelMonitorRadioCard-details"
              class="euiCheckableCard__label"
              for="bucketLevelMonitorRadioCard"
            >
              Per bucket monitor
            </label>
            <div
              class="euiCheckableCard__children"
              id="bucketLevelMonitorRadioCard-details"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding-bottom:10px;padding-top:0px"
              >
                <div
                  class="euiTextColor euiTextColor--subdued"
                >
                  Per bucket monitors run a query that evaluates trigger criteria based on aggregated values in the dataset.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFormRow euiFormRow--hasEmptyLabelSpace"
      id="monitorTypeClusterMetrics-form-row-row"
    >
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder euiPanel--flexGrowZero euiSplitPanel euiSplitPanel--row euiCheckableCard"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--subdued euiPanel--noShadow euiPanel--noBorder euiPanel--flexGrowZero euiPanel--isClickable euiSplitPanel__inner"
          >
            <div
              class="euiRadio euiRadio--noLabel"
              data-test-subj="clusterMetricsMonitorRadioCard"
            >
              <input
                class="euiRadio__input"
                id="clusterMetricsMonitorRadioCard"
                name="monitorTypeClusterMetrics"
                type="radio"
                value="cluster_metrics_monitor"
              />
              <div
                class="euiRadio__circle"
              />
            </div>
          </div>
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--transparent euiPanel--noShadow euiPanel--noBorder euiSplitPanel__inner"
          >
            <label
              aria-describedby="clusterMetricsMonitorRadioCard-details"
              class="euiCheckableCard__label"
              for="clusterMetricsMonitorRadioCard"
            >
              Per cluster metrics monitor
            </label>
            <div
              class="euiCheckableCard__children"
              id="clusterMetricsMonitorRadioCard-details"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding-bottom:10px;padding-top:0px"
              >
                <div
                  class="euiTextColor euiTextColor--subdued"
                >
                  Per cluster metrics monitors run API requests to monitor the clusterâ€™s health.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFormRow euiFormRow--hasEmptyLabelSpace"
      id="monitorTypeDocLevel-form-row-row"
    >
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder euiPanel--flexGrowZero euiSplitPanel euiSplitPanel--row euiCheckableCard"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--subdued euiPanel--noShadow euiPanel--noBorder euiPanel--flexGrowZero euiPanel--isClickable euiSplitPanel__inner"
          >
            <div
              class="euiRadio euiRadio--noLabel"
              data-test-subj="docLevelMonitorRadioCard"
            >
              <input
                class="euiRadio__input"
                id="docLevelMonitorRadioCard"
                name="monitorTypeDocLevel"
                type="radio"
                value="doc_level_monitor"
              />
              <div
                class="euiRadio__circle"
              />
            </div>
          </div>
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--transparent euiPanel--noShadow euiPanel--noBorder euiSplitPanel__inner"
          >
            <label
              aria-describedby="docLevelMonitorRadioCard-details"
              class="euiCheckableCard__label"
              for="docLevelMonitorRadioCard"
            >
              Per document monitor
            </label>
            <div
              class="euiCheckableCard__children"
              id="docLevelMonitorRadioCard-details"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding-bottom:10px;padding-top:0px"
              >
                <div
                  class="euiTextColor euiTextColor--subdued"
                >
                  Per document monitors run queries that return individual documents matching the trigger conditions.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiFlexItem euiFlexItem--flexGrowZero"
  >
    <div
      class="euiFormRow euiFormRow--hasEmptyLabelSpace"
      id="monitorTypeCompositeLevel-form-row-row"
    >
      <div
        class="euiFormRow__fieldWrapper"
      >
        <div
          class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPanel--hasBorder euiPanel--flexGrowZero euiSplitPanel euiSplitPanel--row euiCheckableCard"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--subdued euiPanel--noShadow euiPanel--noBorder euiPanel--flexGrowZero euiPanel--isClickable euiSplitPanel__inner"
          >
            <div
              class="euiRadio euiRadio--noLabel"
              data-test-subj="compositeLevelMonitorRadioCard"
            >
              <input
                class="euiRadio__input"
                id="compositeLevelMonitorRadioCard"
                name="monitorTypeCompositeLevel"
                type="radio"
                value="composite"
              />
              <div
                class="euiRadio__circle"
              />
            </div>
          </div>
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusNone euiPanel--transparent euiPanel--noShadow euiPanel--noBorder euiSplitPanel__inner"
          >
            <label
              aria-describedby="compositeLevelMonitorRadioCard-details"
              class="euiCheckableCard__label"
              for="compositeLevelMonitorRadioCard"
            >
              Composite monitor
            </label>
            <div
              class="euiCheckableCard__children"
              id="compositeLevelMonitorRadioCard-details"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding-bottom:10px;padding-top:0px"
              >
                <div
                  class="euiTextColor euiTextColor--subdued"
                >
                  Composite monitors chain the outputs of different monitor types and focus trigger conditions to reduce alert noise.
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
